### ISSUE TITLE
any way to do causal pad in Flux ?

### ISSUE THREAD

**Issue by neoareslinux** [FluxML/Flux.jl#2505]
### Motivation and description

does Cov support  causal pad ?  I want build causal dilated convolutional network model, but i cannot find any document to guid me to do this with Flux.

**Comment by CarloLucibello** [comment]
Do you have some pytorch reference code? 
I don't see it as an option in pytorch's [Conv1d](https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html)

Maybe just first process the input using [NNlib.pad_constant](https://fluxml.ai/NNlib.jl/stable/reference/#NNlib.pad_constant)?

**Comment by neoareslinux** [comment]
keras code, I have not used pytorch, i learn DL in keras framework and now want to migrate to julia&Flux.

```
def last_time_step_mse(Y_true, Y_pred):
    return keras.metrics.mean_squared_error(Y_true[:, -1], Y_pred[:, -1])

model = keras.models.Sequential()
model.add(keras.layers.InputLayer(input_shape=[None, 1]))
for rate in (1, 2, 4, 8) * 2:
    model.add(keras.layers.Conv1D(filters=20, kernel_size=2, padding="causal",
                                  activation="relu", dilation_rate=rate))
model.add(keras.layers.Conv1D(filters=10, kernel_size=1))
model.compile(loss="mse", optimizer="adam", metrics=[last_time_step_mse])
history = model.fit(X_train, Y_train, epochs=20,
                    validation_data=(X_valid, Y_valid))
```

**Comment by neoareslinux** [comment]
> Do you have some pytorch reference code? I don't see it as an option in pytorch's [Conv1d](https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html)
> 
> Maybe just first process the input using [NNlib.pad_constant](https://fluxml.ai/NNlib.jl/stable/reference/#NNlib.pad_constant)?



> Do you have some pytorch reference code? I don't see it as an option in pytorch's [Conv1d](https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html)
> 
> Maybe just first process the input using [NNlib.pad_constant](https://fluxml.ai/NNlib.jl/stable/reference/#NNlib.pad_constant)?

keras code, I have not used pytorch, i learn DL in keras framework and now want to migrate to julia&Flux.

```
def get_model(input_length, num_features):
    model = keras.Sequential()

    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=1, padding='causal', activation='relu', 
                 input_shape=(input_length, num_features)))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=2, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=4, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=8, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=1, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=2, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=4, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=8, padding='causal', activation='relu'))

    # Flatten the output
    model.add(layers.Flatten())

    # Add a dense layer for binary classification
    model.add(layers.Dense(1, activation='sigmoid'))

    # Compile the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

```

**Comment by neoareslinux** [comment]
> Do you have some pytorch reference code? I don't see it as an option in pytorch's [Conv1d](https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html)
> 
> Maybe just first process the input using [NNlib.pad_constant](https://fluxml.ai/NNlib.jl/stable/reference/#NNlib.pad_constant)?

keras code, I have not used pytorch, i learn DL in keras framework and now want to migrate to julia&Flux.

```
def get_model(input_length, num_features):
    model = keras.Sequential()

    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=1, padding='causal', activation='relu', 
                 input_shape=(input_length, num_features)))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=2, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=4, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=8, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=1, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=2, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=4, padding='causal', activation='relu'))
    model.add(layers.Conv1D(filters=20, kernel_size=2, dilation_rate=8, padding='causal', activation='relu'))

    # Flatten the output
    model.add(layers.Flatten())

    # Add a dense layer for binary classification
    model.add(layers.Dense(1, activation='sigmoid'))

    # Compile the model
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

```

**Comment by CarloLucibello** [comment]
I don't know what Keras does. 
Looking at this
https://gist.github.com/paultsw/7a9d6e3ce7b70e9e2c61bc9287addefc
it seems like you just have to handle the padding appropriately.

**Comment by neoareslinux** [comment]
> I don't know what Keras does. Looking at this https://gist.github.com/paultsw/7a9d6e3ce7b70e9e2c61bc9287addefc it seems like you just have to handle the padding appropriately.

thanks, maybe related  to https://github.com/FluxML/Flux.jl/issues/258

**Comment by CarloLucibello** [comment]
Closing — causal padding is done by padding the input asymmetrically before the convolution, e.g. with `NNlib.pad_zeros` / `NNlib.pad_constant` applied only on the left of the time dimension, then a plain `Conv` with `pad=0`. The gist linked above shows the same approach.

Making `Conv` accept non-zero padding modes directly is tracked in #1917.
